pkg/runtime/framework/plugins/jobset/builder.go

```diff
@@ -101,53 +101,70 @@ func (b *Builder) Initializer(trainJob *trainer.TrainJob) *Builder {
+}
+
 // Trainer updates JobSet values for the trainer Job.
 func (b *Builder) Trainer(info *runtime.Info, trainJob *trainer.TrainJob) *Builder {
 	for i, rJob := range b.Spec.ReplicatedJobs {
 		jobMetadata := rJob.Template.ObjectMetaApplyConfiguration
 		if jobMetadata == nil || jobMetadata.Labels == nil {
 			continue
 		}
+
+		// Update the Parallelism and Completions values for the Trainer Job.
 		if ancestor, ok := jobMetadata.Labels[constants.LabelTrainJobAncestor]; ok && ancestor == constants.AncestorTrainer {
 			// TODO: Support multiple replicas ('.template.spec.replicatedJobs[*].replicas') for replicated Jobs.
 			// REF: https://github.com/kubeflow/trainer/issues/2318
 			b.Spec.ReplicatedJobs[i].Replicas = ptr.To[int32](1)
-			// Update the Parallelism and Completions values for the Trainer Job.
 			b.Spec.ReplicatedJobs[i].Template.Spec.Parallelism = info.FindPodSetByAncestor(constants.AncestorTrainer).Count
 			b.Spec.ReplicatedJobs[i].Template.Spec.Completions = info.FindPodSetByAncestor(constants.AncestorTrainer).Count
+		}
 
-			// Update values for the Trainer container.
-			for j, container := range rJob.Template.Spec.Template.Spec.Containers {
-				if *container.Name == constants.Node {
-					// Update values from the TrainJob trainer.
-					if jobTrainer := trainJob.Spec.Trainer; jobTrainer != nil {
-						if image := jobTrainer.Image; image != nil {
-							b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Image = image
-						}
-						if command := jobTrainer.Command; command != nil {
-							b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Command = command
-						}
-						if args := jobTrainer.Args; args != nil {
-							b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Args = args
-						}
-						if resourcesPerNode := jobTrainer.ResourcesPerNode; resourcesPerNode != nil &&
-							(resourcesPerNode.Limits != nil || resourcesPerNode.Requests != nil) {
-							requirements := corev1ac.ResourceRequirements()
-							if limits := resourcesPerNode.Limits; limits != nil {
-								requirements.WithLimits(limits)
-							}
-							if requests := resourcesPerNode.Requests; requests != nil {
-								requirements.WithRequests(requests)
-							}
-							b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].
-								WithResources(requirements)
-						}
-						apply.UpsertEnvVars(
-							&b.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Env,
-							apply.EnvVars(jobTrainer.Env...)...,
-						)
+		// Apply resources to containers
+		for j, container := range rJob.Template.Spec.Template.Spec.Containers {
+			// Skip if container is neither node nor launcher
+			if *container.Name != constants.Node && *container.Name != constants.Launcher {
+				continue
+			}
+
+			// Skip launcher container if runLauncherAsNode is false
+			if *container.Name == constants.Launcher && !b.isRunLauncherAsNode(info) {
+				continue
+			}
```

Comment:
@tenzen-y @astefanutti How do you like this approach ?
Shall we check that resources need to updated in the JobSet plugin or in the MPI Plugin ?
E.g. we can use the `PodSet` in the Info object.

Reply:
Yes it seems like it would be a better separation of concerns to update the launcher resources in the MPI plugin.

I think the current PR can be merged as is to enable fix the current behavior and follow-up to have a proper design.